src/SUMMARY.md

```diff
@@ -99,20 +104,24 @@
             - [Closure constraints](./borrow_check/region_inference/closure_constraints.md)
             - [Error reporting](./borrow_check/region_inference/error_reporting.md)
         - [Two-phase-borrows](./borrow_check/two_phase_borrows.md)
+    - [Parameter Environments](./param_env.md)
+
+- [Part 5: The Compiler Backend](./part-5-intro.md)
+    - [The MIR (Mid-level IR)](./mir/index.md)
```

Comment:
This also appears in frontend. It also raises question... are operations on MIR, such as optimizations and conversion to llvm-ir considered frontend or backend.

Reply:
"middle end"?

Reply:
@Centril is that a normal compiler term? Also makes me wonder if we should bother with [front,mid,back]-end terms.

Reply:
Honestly, I don't think any of these is a very well defined term.

Backing off of this terminology in favor of something else might be a good idea. eg, we could do "Intermediate Represtations" and "Producing a Binary" as the part titles instead

Reply:
@tshepang "Frontend" and "Backend" are common terms in compilers, but "middle-end" is probably not. Still, it seems fairly intuitive in context?

Reply:
The main problem with the "frontend" and "backend" thing as I see it is that the line between them is fairly arbitrary. One could reasonably say (we don't) that LLVM bitcode generation is still the frontend, and the real backend actually starts inside LLVM.

Reply:
I vote for getting rid of the terms... I don't find their usage useful. Perhaps, if we wanted to split things, we could use more meaningful terms:
```
- parsing
- analysis (includes typeck and borrowck)
- codegen
```
That fees better than
```
- front
- middle
- back
```

Reply:
> The main problem with the "frontend" and "backend" thing as I see it is that the line between them is fairly arbitrary. One could reasonably say (we don't) that LLVM bitcode generation is still the frontend, and the real backend actually starts inside LLVM.

Hmm... IMO, the most intuitive set of terms are "representations".  You have a beginning representation of a program as Rust source text, things happen to it, and you end with a linked binary representation of the program.  There are intermediate representations along the way to get there.  Could we abandon the "frontend", "middle", "backend" terms + associated connotations, and just use "Early intermediate representations", "Late intermediate representations" if we feel that ordering of the representation in the ToC helps one to understand how the compile process works?

I also like the idea of a "You are Here" overview diagram that is used at the beginning of each chapter so that a reader can understand where they are in the full sequence of stages and how that stage relates to adjacent stages.  Maybe this eliminates the need for the order terminology altogether?

Reply:
@chrissimpkins I agree with the focus on IRs :) 👍

Reply:
Ok so how about:
- Source code representations
- Analysis
- From MIR to binary

And I do like @chrissimpkins's idea about a "you are here" map

Reply:
*early* and *late* have same problem with *front* and *back*... one has to choose some (semi-)arbitrary line, and decide what's early and what's late

Reply:
Also, it occurs to me that this goes well with the classic definition of a compiler: a program that takes code in language A and outputs code in language B.